src/macaron/repo_finder/repo_utils.py

```diff
@@ -0,0 +1,279 @@
+# Copyright (c) 2024 - 2024, Oracle and/or its affiliates. All rights reserved.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
+
+"""This module contains the utility functions for repo and commit finder operations."""
+import json
+import logging
+import os
+import string
+
+from git import InvalidGitRepositoryError
+from packageurl import PackageURL
+from pydriller import Git
+
+from macaron.config.global_config import global_config
+from macaron.errors import CloneError, RepoCheckOutError
+from macaron.repo_finder.commit_finder import find_commit
+from macaron.slsa_analyzer.git_service import GIT_SERVICES, BaseGitService
+from macaron.slsa_analyzer.git_service.base_git_service import NoneGitService
+from macaron.slsa_analyzer.git_url import (
+    GIT_REPOS_DIR,
+    check_out_repo_target,
+    get_remote_origin_of_local_repo,
+    get_remote_vcs_url,
+    get_repo_dir_name,
+    is_empty_repo,
+    is_remote_repo,
+    resolve_local_path,
+)
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+def create_filename(purl: PackageURL) -> str:
+    """Create the filename of the report based on the PURL.
+
+    Parameters
+    ----------
+    purl: PackageURL
+        The PackageURL of the artifact.
+
+    Returns
+    -------
+    str
+        The filename to save the report under.
+    """
+
+    def convert_to_path(text: str) -> str:
+        """Convert a PackageURL component to a path safe form."""
+        allowed_chars = string.ascii_letters + string.digits + "-"
+        return "".join(c if c in allowed_chars else "_" for c in text)
+
+    filename = f"{convert_to_path(purl.type)}"
+    if purl.namespace:
+        filename = filename + f"/{convert_to_path(purl.namespace)}"
+    filename = filename + f"/{convert_to_path(purl.name)}/{convert_to_path(purl.name)}.source.json"
+    return filename
+
+
+def generate_report(purl: str, commit: str, repo: str, target_dir: str) -> bool:
+    """Create the report and save it to the passed directory.
+
+    Parameters
+    ----------
+    purl: str
+        The PackageURL of the target artifact, as a string.
+    commit: str
+        The commit hash to report.
+    repo: str
+        The repository to report.
+    target_dir: str
+        The path of the directory where the report will be saved.
+
+    Returns
+    -------
+    bool
+        True if the report was created. False otherwise.
+    """
+    report_json = create_report(purl, commit, repo)
+
+    try:
+        purl_object = PackageURL.from_string(purl)
+    except ValueError as error:
+        logger.debug("Failed to parse purl string as PURL: %s", error)
+        return False
```

Comment:
nit: I think this block can be moved before `report_json = create_report(purl, commit, repo)`